### PR TITLE
Instrument import interface: Sysmex XT4000i

### PR DESCRIPTION
**Pull request by xispa**
## Description of the issue/feature this PR addresses

**IMPORTANT: #536 needs to be accepted first**

Import interface for 'Sysmex XT-4000i' instrument, a fluorescence flow cytometry for high quality haematology analysis. It uses 'Sysmex XT-1800i's Parser (#536), since they are in the same format.
https://www.sysmex-mea.com/products/xt-4000i-919.html

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: bika/lims/exportimport/instruments/resultsimport.py [hunk 1]
```diff
@@ -255,6 +255,70 @@ def _parseline(self, line):
...
+        # We test in import functions if the file was uploaded
+        try:
+            if self._encoding:
+                import codecs
+                f = codecs.open(infile.name, 'r', encoding=self._encoding)
+            else:
+                f = open(infile.name, 'rU')
+        except AttributeError:
+            f = infile
+        for line in f.readlines():
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/resultsimport.py]: This file never gets closed. I'm not sure how this affects Memory in the long run, so I would prefer to use a context manager here, e.g. `with file.open(...)` and read the lines somewhere into a variable. Probably best to do that in a separate method. e.g. `lines = self.read(infile.name)`. There you could also preprocess the lines, e.g. with `lines.strip` or similar

File: bika/lims/exportimport/instruments/resultsimport.py [hunk 2]
```diff
@@ -255,6 +255,70 @@ def _parseline(self, line):
...
+            self._numline += 1
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/resultsimport.py]: Why is this needed?

File: bika/lims/exportimport/instruments/sysmex/xt/i1800.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        """
+        sline = line.split(SEPARATOR)
+        if sline[0] == SEGMENT_HEADER:
+            return self._handle_header(sline)
+        elif sline[0] == SEGMENT_OBSERVATION_ORDER:
+            return self._handle_new_record(sline)
+        elif sline[0] == SEGMENT_RESULT:
+            return self._handle_result_line(sline)
+        elif sline[0] == SEGMENT_EOF:
+            return self._handle_eof(sline)
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/sysmex/xt/i1800.py]: You are accessing here multiple times the first element of the header, therefore I would create a variable like `segment = sline[0]` and then you could think of a mapping of segment -> handler to get it immediately, e.g.
```python
handlers = {
    SEGMENT_HEADER: self._handle_header,
    ...
}
handler = hanlers.get(segment)
...
```

File: bika/lims/exportimport/instruments/sysmex/xt/i4000.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        elif artoapply == 'received_tobeverified':
+            status = ['sample_received', 'attachment_due', 'to_be_verified']
+
+        over = [False, False]
+        if override == 'nooverride':
+            over = [False, False]
+        elif override == 'override':
+            over = [True, False]
+        elif override == 'overrideempty':
+            over = [True, True]
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/sysmex/xt/i4000.py]: create a mapping, e.g.:

```python
override_mapping = {
    "nooverride": [False, False].
    "override": [True, False]
   ...
}
over = override_mapping.get(override, [False, False])
```

File: bika/lims/exportimport/instruments/sysmex/xt/i4000.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        errors.append(t(_("Unrecognized file format ${fileformat}",
+                          mapping={"fileformat": fileformat})))
+
+    if parser:
+        # Load the importer
+        status = ['sample_received', 'attachment_due', 'to_be_verified']
+        if artoapply == 'received':
+            status = ['sample_received']
+        elif artoapply == 'received_tobeverified':
+            status = ['sample_received', 'attachment_due', 'to_be_verified']
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/sysmex/xt/i4000.py]: create a mapping, e.g.:

```python
status_mapping = {
    "received": ['sample_received'],
    "received_tobeverified": ['sample_received', 'attachment_due', 'to_be_verified']
    ...
}
status = status_mapping.get(artoapply, ['sample_received', 'attachment_due', 'to_be_verified'])
```

File: bika/lims/exportimport/instruments/sysmex/xt/i4000.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        elif sample == 'sample_clientsid':
+            sam = ['getSampleID', 'getClientSampleID']
+
+        importer = SysmexXTImporter(parser=parser,
+                                  context=context,
+                                  idsearchcriteria=sam,
+                                  allowed_ar_states=status,
+                                  allowed_analysis_states=None,
+                                  override=over,
+                                  instrument_uid=instrument)
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/sysmex/xt/i4000.py]: Take care of indendation here

File: bika/lims/exportimport/instruments/sysmex/xt/i4000.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        tbex = ''
+        try:
+            importer.process()
+        except:
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/sysmex/xt/i4000.py]: avoid blind excepts

File: bika/lims/exportimport/instruments/sysmex/xt/i1800.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+        elif sample == 'sample_clientsid':
+            sam = ['getSampleID', 'getClientSampleID']
+
+        importer = SysmexXTImporter(parser=parser,
+                                  context=context,
+                                  idsearchcriteria=sam,
+                                  allowed_ar_states=status,
+                                  allowed_analysis_states=None,
+                                  override=over,
+                                  instrument_uid=instrument)
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/sysmex/xt/i1800.py]: Take care of indendation

File: bika/lims/exportimport/instruments/sysmex/xt/i1800.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+        sam = ['getId', 'getSampleID', 'getClientSampleID']
+        if sample == 'requestid':
+            sam = ['getId']
+        if sample == 'sampleid':
+            sam = ['getSampleID']
+        elif sample == 'clientsid':
+            sam = ['getClientSampleID']
+        elif sample == 'sample_clientsid':
+            sam = ['getSampleID', 'getClientSampleID']
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/sysmex/xt/i1800.py]: better use a dictionary mapping here

File: bika/lims/exportimport/instruments/sysmex/xt/i1800.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+        elif artoapply == 'received_tobeverified':
+            status = ['sample_received', 'attachment_due', 'to_be_verified']
+
+        over = [False, False]
+        if override == 'nooverride':
+            over = [False, False]
+        elif override == 'override':
+            over = [True, False]
+        elif override == 'overrideempty':
+            over = [True, True]
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/sysmex/xt/i1800.py]: better use a dictionary mapping here

File: bika/lims/exportimport/instruments/sysmex/xt/i1800.py [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+        errors.append(t(_("Unrecognized file format ${fileformat}",
+                          mapping={"fileformat": fileformat})))
+
+    if parser:
+        # Load the importer
+        status = ['sample_received', 'attachment_due', 'to_be_verified']
+        if artoapply == 'received':
+            status = ['sample_received']
+        elif artoapply == 'received_tobeverified':
+            status = ['sample_received', 'attachment_due', 'to_be_verified']
```
Comment (ramonski) [inline bika/lims/exportimport/instruments/sysmex/xt/i1800.py]: better use a dictionary mapping here

**Review by ramonski — CHANGES_REQUESTED** [review]

File: bika/lims/exportimport/instruments/resultsimport.py [hunk 12]
```diff
@@ -255,6 +255,70 @@ def _parseline(self, line):
...
 
 
+class InstrumentTXTResultsFileParser(InstrumentResultsFileParser):
+
+    def __init__(self, infile, separator, encoding=None,):
+        InstrumentResultsFileParser.__init__(self, infile, 'TXT')
+        # Some Instruments can generate files with different encodings, so we
+        # may need this parameter
+        self._separator = separator
+        self._encoding = encoding
```
[the remark this answers is not in the document]
Comment (xispa) [reply bika/lims/exportimport/instruments/resultsimport.py]: Just to tell the dev that these variables are meant to be private, only for instance's internal use. Just want to make a bit difficult for the dev to do something like `instance._encoding='_wtf'`

File: bika/lims/exportimport/instruments/resultsimport.py (hunk 2 above)
Comment (juangallostra) [reply bika/lims/exportimport/instruments/resultsimport.py]: @ramonski As far as I checked `self._numline` is used to show the line number in the logs

File: bika/lims/exportimport/instruments/resultsimport.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (juangallostra) [reply bika/lims/exportimport/instruments/resultsimport.py]: @ramonski @xispa for now I will leave the underscore.
